Add (optional) styling to fields that are not required

britt-mo opened the pull request:
## Add `(optional)` styling to fields that are not required
[jira ticket](https://qmacbis.atlassian.net/jira/software/c/projects/MDCT/boards/232?modal=detail&selectedIssue=MDCT-2237)

## Summary
Added '(optional)' text to fields. New attribute `styleAsOptional` can be added to mcpar.json to add `(optional)` at the end of the title for a field.

### How to test
Add  `"styleAsOptional": true` in the `props:` object of any field except dynamic fields (should support text, number, radio/checkbox, date, dropdown, single checkbox, etc.)` and check in the browser that the targeted field's title now has `(optional)` at the end of it. This has nothing to do with the field's validation, as that is controlled by `validation` type.

### Important updates
<!-- Any changed dependencies, .env files, local configs, etc. and
instructions for other engineers, e.g. requires new installs in directories -->

### Author checklist
<!-- Complete the following before marking ready for review -->
- [ x] I have performed a self-review of my code
- [ x] I have added [thorough](https://bit.ly/3zPrxuZ) tests, if necessary
- [ ] I have updated the documentation, if necessary

braxex commented on services/ui-src/src/utils/other/styleAsOptional.ts:
```diff
@@ -0,0 +1,4 @@
+import { parseCustomHtml } from "./parsing";
+
+export const labelTextWithOptional = (label: any) =>
```
Request: Strengthen label type to `string`.

braxex commented on services/ui-src/src/utils/other/styleAsOptional.ts:
```diff
@@ -0,0 +1,4 @@
+import { parseCustomHtml } from "./parsing";
```
Request: Let's move this utility method out of `other/styleAsOptional` and put it in `parsing/parsing.ts` since that's what it's doing.

braxex commented on services/ui-src/src/forms/mcpar/mcpar.json:
```diff
@@ -36,7 +36,7 @@
                 "id": "contactName",
                 "type": "text",
-                "validation": "text",
+                "validation": "textOptional",
```
Blocking: Switch this back to `text` because we don't want to change it to optional for real.

britt-mo replied:
good catch

braxex commented on services/ui-src/src/components/fields/ChoiceListField.tsx:
```diff
@@ -226,6 +230,9 @@ const sx = {
   },
+  ".optional-text": {
+    fontWeight: "lighter",
+  },
```
Request: Put this style once in the styling object in `Form.tsx` so it applies to all fields where this class shows up.

braxex replied:
_Note: Applies to all instances below._

braxex commented on services/ui-src/src/components/fields/ChoiceListField.tsx:
```diff
@@ -19,6 +19,7 @@ import {
   FormField,
   InputChangeEvent,
 } from "types";
+import { labelTextWithOptional } from "utils/other/styleAsOptional";
```
Request: Change to export util the same as all other utils through the `utils` dir, then import with the others above on lines 9-12.

braxex replied:
_Note: Applies to all instances below._

braxex commented on services/ui-src/src/components/fields/DropdownField.tsx:
```diff
@@ -152,5 +158,12 @@ interface Props {
   [key: string]: any;
 }
+
+const sx = {
+  ".optional-text": {
+    fontWeight: "lighter",
+  },
+};
```
Request: As noted above, this style should be moved to `Form.tsx` and only be written there once. Then we can remove the `sx` object here. Then you can revert the change on line 134 back to `sx={sxOverride}`.

braxex reviewed (changes requested):
I think we also need to do this work for NumberField and ChoiceField (the single checkbox).

britt-mo commented:
> I think we also need to do this work for NumberField and ChoiceField (the single checkbox).

@braxex I updated ChoiceField but NumberField uses TextField. When I added styleAsOptional to a `type: number`, I saw the '(optional)' text come up. Let me know if I'm not looking at the right thing!

braxex commented on services/ui-src/src/components/fields/DropdownField.tsx:
```diff
@@ -14,7 +14,7 @@ import {
 } from "types";
 import { dropdownDefaultOptionText } from "../../constants";
-import { labelTextWithOptional } from "utils/other/styleAsOptional";
+import { labelTextWithOptional } from "utils/other/parsing";
```
Nit: Can import from `utils` with others on lines 10-14.

braxex commented on services/ui-src/src/utils/other/parsing.test.tsx:
```diff
@@ -58,3 +95,14 @@ describe("Test parseCustomHtml", () => {
   });
 });
+
+describe("Test labelTextWithOptional", () => {
```
Request: Right now this test is pulling in the enter component and report context, etc. and testing the utility where it's used in components. It would be better to move a testing like this to `TextField.test.tsx` where we already have all this mocked out and where we are testing the rest of the component. 

Here in `parsing.test.tsx`, we should just test the `labelTextWithOptional` method itself. We could do that by testing the input and output of the function. So something like: if you pass a label string of "test value", you get back `<span>test value <span class='optional-text'> (optional)</span></span>` or whatever comes back. 

Changing the tests like this helps keep concerns separate. The utility test file just tests the utilities and the component test file just tests the component.

braxex commented on services/ui-src/src/components/fields/TextField.tsx:
```diff
@@ -88,13 +92,11 @@ export const TextField = ({
   const labelClass = !label ? "no-label" : "";
-  const labelText = styleAsOptional ? labelTextWithOptional(label) : label;
+  const labelText =
+    label && styleAsOptional ? labelTextWithOptional(label) : label;
```
Question: As it currently stands, this code will make it so if there's no label, there will also be no "(optional)" on a field. That's probably what we want, but we should double-check with design since that's a bit of an edge case. Will you please reach out to design and see what they want to do if there's no label text? This happens a fair amount with things like nested text boxes under "Other, specify" and similar.

braxex reviewed (changes requested):
Looking good -- just a couple requests.

braxex commented:
> > I think we also need to do this work for NumberField and ChoiceField (the single checkbox).
> 
> @braxex I updated ChoiceField but NumberField uses TextField. When I added styleAsOptional to a `type: number`, I saw the '(optional)' text come up. Let me know if I'm not looking at the right thing!

Ah sure that makes sense. Ok, no worries!

tblackwe commented on services/ui-src/src/components/fields/ChoiceField.tsx:
```diff
@@ -5,12 +5,14 @@ import { Choice as CmsdsChoice } from "@cmsgov/design-system";
   label,
   hint,
   sxOverride,
+  styleAsOptional,
```
I'm sure I'm just missing it, but where is 'styleAsOptional' called? is it part of the report json?

britt-mo replied:
it's an attribute that gets added in the 'props' object in the mcpar.json

britt-mo replied:
It's not getting used yet though so I couldn't add it.

braxex replied:
It will be in MLR, but not yet in MCPAR.
